### PR TITLE
Add proper unipmlemented errors for array copying

### LINKED ISSUE #7635: Improve error messages for unsupported calldata arrays in old code generation.
Given that we closed https://github.com/ethereum/solidity/pull/6983 and https://github.com/ethereum/solidity/pull/6881 and decided not to actually fix https://github.com/ethereum/solidity/issues/6835, we should at least provide a decent error message indicating possible workarounds for the cases that won't be implemented in old code generation.

### PR DESCRIPTION
**Pull request by Marenz**
fixes #7635

### CODE REVIEW

File: test/libsolidity/syntaxTests/array/nested_calldata_memory.sol [hunk 1]
```diff
@@ -0,0 +1,15 @@
+pragma experimental ABIEncoderV2;
+
+contract Test {
+    struct shouldBug {
+        bytes[2] deadly;
+    }
+    mapping(uint256 => shouldBug) graveyard;
```
Comment (hrkrshnn) [inline test/libsolidity/syntaxTests/array/nested_calldata_memory.sol]: Is the mapping needed here?

File: libsolidity/codegen/CompilerUtils.cpp [hunk 2]
```diff
@@ -974,6 +974,9 @@ void CompilerUtils::convertType(
 				}
 				else
 				{
+					if (auto baseType = dynamic_cast<ArrayType const*>(typeOnStack.baseType()))
+						solUnimplementedAssert(typeOnStack.isDynamicallySized() || !typeOnStack.isDynamicallyEncoded() || !baseType->isByteArray(), "Copying nested calldata dynamic arrays from calldata to memory is not implemented in the old codegen.");
+
```
Comment (hrkrshnn) [inline libsolidity/codegen/CompilerUtils.cpp]: Is it possible to add location here?

File: test/libsolidity/syntaxTests/array/nested_calldata_memory.sol (hunk 1 above)
Comment (Marenz) [reply test/libsolidity/syntaxTests/array/nested_calldata_memory.sol]: Good point, removed

File: libsolidity/codegen/CompilerUtils.cpp (hunk 2 above)
Comment (Marenz) [reply libsolidity/codegen/CompilerUtils.cpp]: to the assert? No, those are more lower-level errors, not proper error reportings.

File: Changelog.md [hunk 3]
```diff
@@ -10,6 +10,7 @@ Compiler Features:
 
 Bugfixes:
  * Optimizer: Fixed a bug in BlockDeDuplicator.
+ * Code Generator: Trigger proper unipmlemented errors instead of generic stack disturbance on certain array copy operations
```
Comment (chriseth) [inline Changelog.md]:
```suggestion
 * Code Generator: Trigger proper unimplemented errors on certain array copy operations.
```

File: libsolidity/codegen/ArrayUtils.cpp [hunk 4]
```diff
@@ -183,6 +183,13 @@ void ArrayUtils::copyArrayToStorage(ArrayType const& _targetType, ArrayType cons
 			// copy
 			if (sourceBaseType->category() == Type::Category::Array)
 			{
+				solUnimplementedAssert(
+					_sourceType.location() != DataLocation::CallData ||
+					_sourceType.isDynamicallySized() ||
```
Comment (chriseth) [inline libsolidity/codegen/ArrayUtils.cpp]: Are you sure this should not be negated?

File: libsolidity/codegen/CompilerUtils.cpp [hunk 5]
```diff
@@ -974,6 +974,9 @@ void CompilerUtils::convertType(
 				}
 				else
 				{
+					if (auto baseType = dynamic_cast<ArrayType const*>(typeOnStack.baseType()))
+						solUnimplementedAssert(typeOnStack.isDynamicallySized() || !typeOnStack.isDynamicallyEncoded() || !baseType->isByteArray(), "Copying nested calldata dynamic arrays from calldata to memory is not implemented in the old codegen.");
```
Comment (chriseth) [inline libsolidity/codegen/CompilerUtils.cpp]: line break.

File: libsolidity/codegen/ArrayUtils.cpp (hunk 4 above)
Comment (Marenz) [reply libsolidity/codegen/ArrayUtils.cpp]: The type that makes problems is `uint[][2]`, so the outer array is a non-dynamically sized one

Comment (chriseth) [reply libsolidity/codegen/ArrayUtils.cpp]: But `uint[][]` would not work either, would it?

File: libsolidity/codegen/ArrayUtils.cpp [hunk 6]
```diff
@@ -183,6 +183,13 @@ void ArrayUtils::copyArrayToStorage(ArrayType const& _targetType, ArrayType cons
...
+					!_sourceType.isDynamicallyEncoded()
+					,
```
Comment (chriseth) [inline libsolidity/codegen/ArrayUtils.cpp]:
```suggestion
					!_sourceType.isDynamicallyEncoded(),
```

File: libsolidity/codegen/ArrayUtils.cpp (hunk 4 above)
Comment (Marenz) [reply libsolidity/codegen/ArrayUtils.cpp]: `uint[][]` is dynamically sized, so that does work

File: libsolidity/codegen/CompilerUtils.cpp [hunk 7]
```diff
@@ -974,6 +974,14 @@ void CompilerUtils::convertType(
 				}
 				else
 				{
+					if (auto baseType = dynamic_cast<ArrayType const*>(typeOnStack.baseType()))
+						solUnimplementedAssert(
+							typeOnStack.isDynamicallySized() ||
+							!typeOnStack.isDynamicallyEncoded() ||
+							!baseType->isByteArray(),
+							"Copying nested calldata dynamic arrays from calldata to memory is not implemented in the old codegen."
```
Comment (chriseth) [inline libsolidity/codegen/CompilerUtils.cpp]:
```suggestion
							"Copying nested dynamic calldata arrays from calldata to memory is not implemented in the old code generator."
```

File: libsolidity/codegen/ArrayUtils.cpp [hunk 8]
```diff
@@ -185,6 +185,13 @@ void ArrayUtils::copyArrayToStorage(ArrayType const& _targetType, ArrayType cons
 			{
 				solAssert(byteOffsetSize == 0, "Byte offset for array as base type.");
 				auto const& sourceBaseArrayType = dynamic_cast<ArrayType const&>(*sourceBaseType);
+
+				solUnimplementedAssert(
+					_sourceType.location() != DataLocation::CallData ||
+					!_sourceType.isDynamicallyEncoded() ||
+					!sourceBaseArrayType.isDynamicallySized(),
+					"Copying nested calldata dynamic arrays from calldata to storage is not implemented in the old codegen."
```
Comment (chriseth) [inline libsolidity/codegen/ArrayUtils.cpp]:
```suggestion
					"Copying nested dynamic calldata arrays from calldata to storage is not implemented in the old codegen."
```

File: libsolidity/codegen/CompilerUtils.cpp [hunk 9]
```diff
@@ -974,6 +974,14 @@ void CompilerUtils::convertType(
 				}
 				else
 				{
+					if (auto baseType = dynamic_cast<ArrayType const*>(typeOnStack.baseType()))
+						solUnimplementedAssert(
+							typeOnStack.isDynamicallySized() ||
+							!typeOnStack.isDynamicallyEncoded() ||
+							!baseType->isByteArray(),
```
Comment (chriseth) [inline libsolidity/codegen/CompilerUtils.cpp]: Why is this assertion different than the other?

Comment (Marenz) [reply libsolidity/codegen/CompilerUtils.cpp]: I used the test cases to tailor the asserts to them

Comment (Marenz) [reply libsolidity/codegen/CompilerUtils.cpp]: I basically lacked the understanding of the bigger picture.

Comment (Marenz) [reply libsolidity/codegen/CompilerUtils.cpp]: I am now extending the assert

**Review by chriseth — APPROVED** [review]

**Pull request merged**